diff -r 6cf9045d4e50 modules/mod_admin/actions/action_admin_delete_media.erl
--- a/modules/mod_admin/actions/action_admin_delete_media.erl	Sat Feb 06 08:24:42 2010 +0100
+++ b/modules/mod_admin/actions/action_admin_delete_media.erl	Sat Feb 06 08:42:28 2010 +0100
@@ -39,7 +39,7 @@
 %% @doc Delete a media.  After the deletion the user is redirected, and/or some items on the page are faded out.
 %% @spec event(Event, Context1) -> Context2
 event({postback, {delete_media, Id, OnSuccess}, _TriggerId, _TargetId}, Context) ->
-    case z_acl:rsc_editable(Id, Context) of
+    case z_acl:rsc_deletable(Id, Context) of
         true ->
             ok = m_media:delete(Id, Context),
             lists:foldl(
diff -r 6cf9045d4e50 modules/mod_admin/actions/action_admin_delete_rsc.erl
--- a/modules/mod_admin/actions/action_admin_delete_rsc.erl	Sat Feb 06 08:24:42 2010 +0100
+++ b/modules/mod_admin/actions/action_admin_delete_rsc.erl	Sat Feb 06 08:42:28 2010 +0100
@@ -39,7 +39,7 @@
 %% @doc Delete a resource.  After the deletion the user is redirected, and/or some items on the page are faded out.
 %% @spec event(Event, Context1) -> Context2
 event({postback, {delete_rsc, Id, OnSuccess}, _TriggerId, _TargetId}, Context) ->
-    case z_acl:rsc_editable(Id, Context) of
+    case z_acl:rsc_deletable(Id, Context) of
         true ->
             ok = m_rsc:delete(Id, Context),
             lists:foldl(
diff -r 6cf9045d4e50 modules/mod_admin/actions/action_admin_dialog_delete_media.erl
--- a/modules/mod_admin/actions/action_admin_dialog_delete_media.erl	Sat Feb 06 08:24:42 2010 +0100
+++ b/modules/mod_admin/actions/action_admin_dialog_delete_media.erl	Sat Feb 06 08:42:28 2010 +0100
@@ -39,7 +39,7 @@
 %% @doc Fill the dialog with the delete confirmation template. The next step will ask to delete the media
 %% @spec event(Event, Context1) -> Context2
 event({postback, {delete_media_dialog, Id, OnSuccess}, _TriggerId, _TargetId}, Context) ->
-    case z_acl:media_editable(Id, Context) of
+    case z_acl:media_deletable(Id, Context) of
         true ->
             Vars = [
                 {on_success, OnSuccess},
diff -r 6cf9045d4e50 modules/mod_admin/actions/action_admin_dialog_delete_rsc.erl
--- a/modules/mod_admin/actions/action_admin_dialog_delete_rsc.erl	Sat Feb 06 08:24:42 2010 +0100
+++ b/modules/mod_admin/actions/action_admin_dialog_delete_rsc.erl	Sat Feb 06 08:42:28 2010 +0100
@@ -39,7 +39,7 @@
 %% @doc Fill the dialog with the delete confirmation template. The next step will ask to delete the resource
 %% @spec event(Event, Context1) -> Context2
 event({postback, {delete_rsc_dialog, Id, OnSuccess}, _TriggerId, _TargetId}, Context) ->
-    case z_acl:rsc_editable(Id, Context) of
+    case z_acl:rsc_deletable(Id, Context) of
         true ->
             Vars = [
                 {on_success, OnSuccess},
diff -r 6cf9045d4e50 modules/mod_admin/templates/admin_edit.tpl
--- a/modules/mod_admin/templates/admin_edit.tpl	Sat Feb 06 08:24:42 2010 +0100
+++ b/modules/mod_admin/templates/admin_edit.tpl	Sat Feb 06 08:42:28 2010 +0100
@@ -340,7 +340,7 @@
 									</div>
 
 									<div class="form-item clearfix">
-										{% button class="discard-resource do_tooltip" disabled=r.is_protected|ornot:is_editable id="delete-button" text="delete" action={dialog_delete_rsc id=r.id on_success={redirect back}} title="Delete this page." %}
+										{% button class="discard-resource do_tooltip" disabled=r.is_protected|ornot:m.rsc[id].is_deletable id="delete-button" text="delete" action={dialog_delete_rsc id=r.id on_success={redirect back}} title="Delete this page." %}
 
 										{% if is_editable %}
 											{% button type="submit" id="save_duplicate" class="save-resource do_tooltip" text="duplicate" title="Duplicate this page." %}
diff -r 6cf9045d4e50 src/models/m_rsc.erl
--- a/src/models/m_rsc.erl	Sat Feb 06 08:24:42 2010 +0100
+++ b/src/models/m_rsc.erl	Sat Feb 06 08:42:28 2010 +0100
@@ -45,7 +45,7 @@
     
 	exists/2, 
 	
-	is_visible/2, is_editable/2, is_ingroup/2, is_me/2, 
+	is_visible/2, is_editable/2, is_deletable/2, is_ingroup/2, is_me/2, 
 	is_cat/3,
 	is_a/2,
 	is_a_id/2,
@@ -279,6 +279,14 @@
             false
     end.
     
+is_deletable(Id, Context) -> 
+    case rid(Id, Context) of
+        RscId when is_integer(RscId) ->
+            z_acl:rsc_deletable(RscId, Context);
+        _ ->
+            false
+    end.
+    
 is_ingroup(Id, Context) -> 
     case rid(Id, Context) of
         RscId when is_integer(RscId) ->
@@ -327,6 +335,7 @@
 p_no_acl(Id, is_me, Context) -> is_me(Id, Context);
 p_no_acl(Id, is_visible, Context) -> is_visible(Id, Context);
 p_no_acl(Id, is_editable, Context) -> is_editable(Id, Context);
+p_no_acl(Id, is_deletable, Context) -> is_deletable(Id, Context);
 p_no_acl(Id, is_ingroup, Context) -> is_ingroup(Id, Context);
 p_no_acl(Id, is_a, Context) -> [ {C,true} || C <- is_a(Id, Context) ];
 p_no_acl(Id, exists, Context) -> exists(Id, Context);
diff -r 6cf9045d4e50 src/models/m_rsc_update.erl
--- a/src/models/m_rsc_update.erl	Sat Feb 06 08:24:42 2010 +0100
+++ b/src/models/m_rsc_update.erl	Sat Feb 06 08:42:28 2010 +0100
@@ -53,7 +53,7 @@
 %% @doc Delete a resource
 %% @spec delete(Props, Context) -> ok | {error, Reason}
 delete(Id, Context) when is_integer(Id), Id /= 1 ->
-    case z_acl:rsc_editable(Id, Context) of
+    case z_acl:rsc_deletable(Id, Context) of
         true ->
             case m_rsc:is_a(Id, category, Context) of
                 true ->
diff -r 6cf9045d4e50 src/support/z_acl.erl
--- a/src/support/z_acl.erl	Sat Feb 06 08:24:42 2010 +0100
+++ b/src/support/z_acl.erl	Sat Feb 06 08:42:28 2010 +0100
@@ -28,6 +28,7 @@
 -export([
     rsc_visible/2,
     rsc_editable/2,
+    rsc_deletable/2,
     rsc_ingroup/2,
     group_editable/2,
     sudo/1,
@@ -96,6 +97,14 @@
 rsc_editable(Id, Context) ->
     acl_editable(m_rsc:get_acl_props(Id, Context), Context).
 
+%% @doc Check if the resource is deletable by the current user
+rsc_deletable(_Id, #context{user_id=undefined}) ->
+    % Anonymous visitors can't delete anything
+    false;
+rsc_deletable(Id, Context) ->
+    acl_deletable(m_rsc:get_acl_props(Id, Context), Context).
+
+
 %% @doc Check if the group of the resource is in one of the groups of the user
 rsc_ingroup(_Id, #context{user_id=undefined}) ->
     false;
@@ -449,3 +458,21 @@
                     lists:member(Acl#acl_props.group_id, Gs)
             end
     end.
+
+
+%% @doc Check if something with the given #acl fields is editable.
+%% @spec acl_editable(#acl_props, #context) -> bool()
+acl_deletable(Acl, Context) ->
+    case Acl#acl_props.is_authoritative of
+        true -> 
+            acl_editable(Acl, Context);
+        false ->
+            case has_role(admin, Context) of
+                true ->
+                    true;
+                false ->
+                                                % Must be in one of our groups
+                    Gs = groups_member(Context),
+                    lists:member(Acl#acl_props.group_id, Gs)
+            end
+    end.
